### DISCUSSION TITLE
Webhook missing data

### DISCUSSION THREAD

**Discussion by jxs714** [pretix/pretix#3661] [Q&A]
Hey guys,

Just tried the webhook setup. Upon testing, only the following information is sent:

{
  "notification_id": 371,
  "organizer": "organizernamehere",
  "customer": "MY3XXS8",
  "action": "pretix.customer.changed"
}

The customer name, address, and custom questions are missing from the hook. Tried multiple hooks including new customer account and customer order, and the results are the same. Am I required to manually modify pretix system files to include more information in the webhook data sent? If so, what files and area? Also, what would the fields be for the default and custom questions? There isn't much information in the documentation.

The goal is to send customer data to a webhook connection application on Wordpress that sends the data to mailster, a mailing list software. Id like to automatically add all customer order data such as name, address and custom questions to a customer mailing list.

Please advise, thanks!

**Comment by pc-coholic** [comment]
Since webhooks are sent async, pwe do not include the data of the object. By the time you get the webhook, it might have changed already again...

The idea here is, that the receiving application should query the required information using the pretix API.

**Reply by jxs714** [reply]
Hi @pc-coholic 

Thanks for the response and update. I was able to figure out setting the Pretix API with a webhook and was able to get all customer information by using Token API for authentication. I decided to go with n8n software, which is a workflow tool similar to zapier. At the last step of a http request node, it sends the customer data to Acelle Mail software and adds them to the mailing list successfully.

I have a question in regards to getting consistent data. Everything is connecting great so far, and I am able to get the orders for an event with the following API End point

https://tickets.domain.com/api/v1/organizers/ORGANIZERNAME/orders/

What I am doing is getting the following customer data:

Node 1
• Name
• Address
• City
• State
• Zipcode

Node 2
• Ticket Type

The nodes combine and get all of the data mentioned from customers. My only issue now is every time I call this endpoint, ALL orders show up in the scheme/tables and it seems impractical as the data exports all orders instead of a recent one. Each order has has its own ID, so if I enter no ID, it will only copy specific data for a single customer order that I entered as JSON. I am trying to capture all of the data above every time that a new order comes in, to then automatically add each correct attendee information to the mailing list with the correct data of that customer.

For example, this is what is currently happening. 

Order ID 1 is

John Doe
111 Address Avenue
New York
NY
11111

Order ID 2 is

Jane Doe
1245 Address Avenue
Chicago
IL
92938

If I enter the json variable {{ $json.results[1].positions[0].attendee_name }} - It will equal to John Doe. When it runs the webhook it will snatch that same exact customers data (John Doe) each time and add them to the mailing list because im using the same json with the results being "json.results[1]" in that json code, which signifies order ID1

Is there a specific endpoint, specific variables or a specific configuration required for me to do this correctly? Am I using the wrong approach?

**Reply by raphaelm** [reply]
Why do you not use the endpoint to get a specific order?

tickets.domain.com/api/v1/organizers/ORGANIZERNAME/orders/CODE/

You get the CODE in the webhook

**Reply by jxs714** [reply]
Hi Raphael

Thanks for the response! Hope you and the team are doing well. 

Is there a variable for CODE at the end of the url so that the webhook can capture and make a unique entry for each customer as they purchase a ticket when the node sends over to the customer information to the mailing list software?

-----------------------------------------
Here is the main goal from a - z:
- Customer 1 purchases general admission ticket
- Pretix API keys serves the customer data (name, address, etc, ticket type) to n8n
- n8n formats and combines the data and sends to acelle mail (mailing software api) and signs the customer up to the general admission mailing list.

- Customer 2 purchases general admission ticket at the same time or after, but ultimately Pretix knows to give them a unique result[ID]
------------------------------------------

I am trying to get each order to be captured with the 1 full customer info + ticket type then sign them up to the mailing list software. The thing is the order ID JSON result command will be different each time for each customer, and it doesnt seem to allow me to put a asterisk like the following to make it auto propagate any new orders with a unique order ID.

Sample
{{ $json.results[*].positions[0].attendee_name }}

I tried this and it didn't work as it would not accept the *. When I use the /orders api page, it displays every single order in the system and I would need to create a separate json result filter for each customer order every time they come in

ie: I create a filter with {{ $json.results[25].positions[0].attendee_name }}
then create another filter for {{ $json.results[26].positions[0].attendee_name }} when the next order comes in. I would have to make a numeric result manually each time.  The aim for me is to capture ONLY the newest customer order, add them to the list, and repeat for each one that comes after.

If I enter the CODE manually, it will snatch that 1 customer data, but that would then be counter productive as I would be manually creating a unique request for each customer order code URL (ie: tickets.domain.com/api/v1/organizers/ORGANIZERNAME/orders/SAMPLECODEHERE/ through the api. Which means I would then be manually entering each ticket order code and new customer data just to add them to the mailing list. The aim is for automation where I do not have to end up exporting all customers order information in pretix to then upload the CSV to the mailing software to a specified mailing list (ie General Admission, VIP etc).

Another thing is using the /order API endpoint displays ALL orders, which means if we have 1,000's of orders, the webhook may overload and overuse resources each time just to find a new customer that hasn't been added to the mailing list. The webhook needs to be setup up with pretix automating and informing the webhook basically talking to n8n saying "Hey, here is a new order and customer, add this to the mailing list" without having to load ALL orders with the /order API endpoint. I couldn't find a specific endpoint that gets new data only or does what I am aiming for. Not sure which is the correct route to go. (Sorry for the long message)

Hope that makes sense or maybe I am missing something

**Reply by raphaelm** [reply]
I'm not sure what the specific question here is. 

> Is there a variable for CODE at the end of the url so that the webhook can capture and make a unique entry for each customer as they purchase a ticket when the node sends over to the customer information to the mailing list software?

What do you mean by "at the end of the URL"? The code is supplied in the body of the webhook.

**Reply by jxs714** [reply]
Hi Raphael,

Thanks for the reply. The order code wouldn't suffice as an endpoint to enter for the webhook as each customer has their own. 

https://tickets.domain.com/api/v1/organizers/ORGANIZERNAME/orders/ <--- this gives every customer order information and would unnecessarily overload the system if there are 1,000s of tickets sold

tickets.domain.com/api/v1/organizers/ORGANIZERNAME/orders/ORDERCODE/ <--- This gives me one unique customer order, and I dont receive the final URL until after the order is complete. This order code information, along with the customer order information, appears in the previous URL mentioned when running as endpoint.

These routes will not allow me to capture and organize the customer data to prepare them to sign them up to mailing list. Each customer order attendee name has a unique json command. For example: {{ $json.results[25].positions[0].attendee_name }} <-- this json results[25] is the 25th customer order. I would only want to display and capture new orders, not resend results[0] - results[25] and have a bunch of all customer data from the event.

After reviewing other options, I think the best route is to use the endpoint api's for exporting the customer order data as the endpoints mentioned in this thread requires a more intricate process.

**Different route: Export Data API's
Here are the steps I am aiming for as the alternative**
- Initiate customer orders export via API by date. For my end, I am aiming for exporting every 12 hrs. 
- The export will give me all full order information (all orders within a 12 hr span) to prepare to import fresh data to mailing list software. I will have to route the cells from the excel to their correction data entries from there.

What endpoint URL structure should I be using to process this correctly?

**Reply by raphaelm** [reply]
I will probably no longer reply here, as it gets lengthy and confusing and I have the impression that it will not lead anywhere.

You seem to have a *significant* misunderstanding about how webhooks work. 

> The order code wouldn't suffice as an endpoint to enter for the webhook as each customer has their own.

Not as an endpoint for the webhook. The webhook however **gets** the order code **as part of the body** and can then request just that order code from the API.

What you are trying to do will be chaotic and prone to fail due to race conditions etc.
